File: practice/cpp/dynamic_programming/stone_game.cpp
// problem: https://leetcode.com/problems/stone-game/

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

bool stoneGame(vector<int> piles)
{
    int n = piles.size();
    vector<vector<int>> dp(n+1, vector<int>(n+1,0));

    for (int d = 1; d <= n; d++) {
        for (int i = 0, j = d-1; j < n; i++,j++) {
            if (d % 2) {
                dp[i][j] = min(-piles[i]+dp[i+1][j], -piles[i]+ (j>0 ? dp[i][j-1] : 0) );
            } else {
                dp[i][j] = max(piles[i]+dp[i+1][j], piles[i]+ (j>0 ? dp[i][j-1] : 0) );
            }
        }
    }

    return dp[0][n-1] > 0;

}

int main(void)
{
    cout << stoneGame({5,4,3,5}) << endl;
}

File: practice/python/dynamic_programming/stone_game.py
"""
- Problem: https://leetcode.com/problems/stone-game/
"""

def stoneGame(piles: list):
    n = len(piles)    
    dp = [[0] * n for i in range(n)]
    for i in range(n):
        dp[i][i] = piles[i]

    for d in range(1,n):
        for i in range(n - d):
            dp[i][i+d] = max(piles[i] - dp[i+1][i+d], piles[i+d] - dp[i][i+d-1])
    return dp[0][-1] > 0

if __name__ == "__main__":
    piles = [7,6,1,5,9,9]
    print(stoneGame(piles))

